src/process_utility.c

```diff
@@ -1688,6 +1687,33 @@ process_altertable_end_subcmd(Hypertable *ht, Node *parsetree, ObjectAddress *ob
 		case AT_DropCluster:
 			foreach_chunk(ht, process_altertable_chunk, cmd);
 			break;
+		case AT_SetTableSpace:
+			{
+				Oid			tspc_oid = get_rel_tablespace(ht->main_table_relid);
+				NameData	tspc_name;
+				Tablespaces *tspcs;
+
+				namestrcpy(&tspc_name, get_tablespace_name(tspc_oid));
+				Assert(OidIsValid(tspc_oid));
+
+				/*
+				 * this is called after postgres ALTER TABLE so command
+				 * tablespace name
+				 */
+				/* is same as current relation tablespace name */
+				Assert(strcmp(cmd->name, NameStr(tspc_name)) == 0);
+				tspcs = tablespace_scan(ht->fd.id);
+				if (tspcs->num_tablespaces > 1)
+					ereport(ERROR,
+							(errcode(ERRCODE_IO_OPERATION_NOT_SUPPORTED),
+							 errmsg("cannot ALTER TABLE SET TABLESPACE when multiple tablespaces are attached to hypertable %s",
+									get_rel_name(ht->main_table_relid)),
+							 errhint("Detach some/all tablespaces before altering the hypertable")));
+
+				tablespace_attach_internal(&tspc_name, ht->main_table_relid, true);
```

Comment:
Would it not be better to have `tablespace_attach_internal` and similar internal functions take `oid` instead of name?
It internally converts oid back to name and here we do the opposite conversion which looks not ideal.
The only function for which it is easier to pass a name is user exposed attach_tablespace function which can just do the conversion inside.

Reply:
I think you are right. but do this optionally at the end. If you have time